File: unity/com.huggingbutt.tiptoestep/Runtime/Messenger.cs
using System;
using System.IO;
using System.Net.Sockets;

// message protocol flag
// 0x01 produced by C#
// 0x02 consumed by C#
// 0x03 produced by Python
// 0x04 consumed by Python
// 0x05 event flag
// 0x06 writing by C#
// 0x07 writing by Python


namespace tiptoestep
{
    public class Messenger : IDisposable
    {
        private long size;
        private TcpClient client;
        private NetworkStream stream;
        private const int MAX_FRAME_BYTES = 10 * 1024 * 1024; // 10MB safety cap

        public Messenger(string host, string port, long size = 1024 * 16)
        {
            this.size = size;
            try
            {
                client = new TcpClient(host, int.Parse(port));
                client.NoDelay = true; // Disable Nagle for low-latency small packets
                stream = client.GetStream();
                // Optionally set timeouts (can be adjusted if needed)
                try { stream.ReadTimeout = 30000; stream.WriteTimeout = 30000; } catch {}
            }
            catch (ArgumentNullException e)
            {
                Console.WriteLine($"ArgumentNullException: {e}");
            }
            catch (SocketException e)
            {
                Console.WriteLine($"SocketException: {e}");
            }
        }

        public void Send(Message msg)
        {
            using (var stream = new MemoryStream())
            {
                using (var writer = new BinaryWriter(stream))
                {
                    writer.Write((byte)0x06); ; // Set writing flag
                    byte[] msgBytes = MessageSerializer.Serialize(msg);
                    writer.Write(BitConverter.GetBytes(msgBytes.Length));
                    
                    writer.Write(msgBytes);
                    writer.Write((byte)0x01); // Set write over flag.
                }
                byte[] send_data = stream.ToArray();
                this.stream.Write(send_data, 0, send_data.Length);
            }

        }

        private byte[] ReadExact(int n)
        {
            byte[] buffer = new byte[n];
            int offset = 0;
            while (offset < n)
            {
                int read = this.stream.Read(buffer, offset, n - offset);
                if (read <= 0)
                {
                    throw new IOException("Socket closed while reading data");
                }
                offset += read;
            }
            return buffer;
        }

        public Message Receive()
        {
            byte[] flag = ReadExact(1);
            byte[] dataLengthBytes = ReadExact(4);
            int dataLength = BitConverter.ToInt32(dataLengthBytes, 0);
            if (dataLength < 0 || dataLength > MAX_FRAME_BYTES)
            {
                throw new InvalidDataException($"Payload length out of bounds: {dataLength}");
            }
            byte[] msgBytes = ReadExact(dataLength);
            byte[] overFlag = ReadExact(1);

            Message msg = null;
            if (flag[0] != 0x07 || overFlag[0] != 0x03) throw new Exception("Incorrect data format received from Python.");
            msg = MessageSerializer.Deserialize(msgBytes);
            return msg;
        }

        public bool Check() // Check if the action message has arrived.
        {
            return this.stream.DataAvailable;
        }

        public void SendReady(uint env_id, uint pid)
        {
            Message msg = new Message();
            msg.pid = pid;
            msg.env_id = env_id;
            msg.agent_id = 0;
            msg.step_id = 0;
            msg.obs_frame_id = 0;
            msg.act_frame_id = 0;
            msg.silent = true;
            msg.message_type = MessageType.Control;
            msg.step_type = StepType.Ready;
            this.Send(msg);
        }

        public void SendEnd(uint env_id, uint pid)
        {
            Message msg = new Message();
            msg.pid = pid;
            msg.env_id = env_id;
            msg.agent_id = 0;
            msg.step_id = 0;
            msg.obs_frame_id = 0;
            msg.act_frame_id = 0;
            msg.silent = true;
            msg.message_type = MessageType.Control;
            msg.step_type = StepType.End;
            this.Send(msg);
        }


        public void Dispose()
        {
            try { this.stream?.Dispose(); } catch {}
            try { this.stream?.Close(); } catch {}
            try { this.client?.Dispose(); } catch {}
            try { this.client?.Close(); } catch {}
        }
    }
}

File: python/tiptoestep/messenger.py
import socket
import struct
import select
import numpy as np
from .proto import Message, MessageSerializer, MessageType, StepType
from typing import Optional


class Messenger:
    def __init__(self, pid=0, env_id=0, host=None, port=None, timeout: Optional[float] = None):
        self.pid = pid
        self.env_id = env_id
        self.server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        # Reuse address to reduce TIME_WAIT issues during rapid restarts
        self.server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self.port: int = port if port else 10086 + pid
        self.host: str = host if host else '127.0.0.1'
        self.timeout = timeout
        self.server_socket.bind((self.host, self.port))
        self.server_socket.listen(5)
        self.client_socket: socket.socket = None

    def listen(self):
        # print(f"Server is listening on {self.host}:{self.port}")
        self.client_socket, addr = self.server_socket.accept()
        # Disable Nagle to reduce latency for small control packets
        try:
            self.client_socket.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
        except Exception:
            pass
        if self.timeout is not None:
            try:
                self.client_socket.settimeout(self.timeout)
            except Exception:
                pass
        # print(f"Connection from {addr}")

    def send(self, msg: Message):
        data = struct.pack('c', b'\x07')
        msg_bytes = MessageSerializer.serialize(msg)
        # Use little-endian explicit packing for length
        data += struct.pack('<I', int(len(msg_bytes)))
        data += msg_bytes
        data += struct.pack('c', b'\x03')
        # Ensure full write over TCP
        self.client_socket.sendall(data)

    def _read_exact(self, n: int) -> bytes:
        """Read exactly n bytes from the client socket or raise if connection closes early."""
        buf = bytearray()
        while len(buf) < n:
            chunk = self.client_socket.recv(n - len(buf))
            if not chunk:
                raise RuntimeError("Socket closed while reading data")
            buf += chunk
        return bytes(buf)

    def send_action(self, agent_id, step_id, action, cmds=None):
        msg = Message()
        msg.pid = self.pid
        msg.env_id = self.env_id
        msg.agent_id = agent_id
        msg.step_id = step_id
        msg.obs_frame_id = 0
        msg.act_frame_id = 0
        msg.silent = False
        msg.message_type = MessageType.Action.value
        msg.step_type = StepType.Step.value
        msg.action = action
        msg.cmds = cmds
        self.send(msg)

    def send_control(self, agent_id, step_id, signal, cmds=None):
        if signal not in ['reset', 'end', 'ready']:
            raise RuntimeError(f"Unable to recognize {signal} control signal.")

        msg = Message()
        msg.pid = self.pid
        msg.env_id = self.env_id
        msg.agent_id = agent_id
        msg.step_id = step_id
        msg.obs_frame_id = 0
        msg.act_frame_id = 0
        msg.silent = False
        msg.message_type = MessageType.Control.value

        if signal == 'reset':
            msg.step_type = StepType.Reset.value
        elif signal == 'end':
            msg.step_type = StepType.End.value
        elif signal == 'ready':
            msg.step_type = StepType.Ready.value

        msg.cmds = cmds
        self.send(msg)

    def check(self):
        """
        Check if there is data available to read from the socket.
        :return:
        """
        readable, _, _ = select.select([self.client_socket], [], [], 0)
        return bool(readable)

    def receive(self, check_obs=False):
        # Read framed packet: [0x06][len(4 LE)][payload][0x01]
        flag = self._read_exact(1)
        data_length_bytes = self._read_exact(4)
        data_length = struct.unpack('<I', data_length_bytes)[0]
        msg_bytes = self._read_exact(data_length)
        over_flag = self._read_exact(1)

        assert flag[0] == 0x06 and over_flag[0] == 0x01, "Incorrect data format received from C#."
        msg = MessageSerializer.deserialize(msg_bytes)

        if check_obs:
            if not (msg.message_type == MessageType.Observation.value
                    and msg.step_type == StepType.Step.value):
                raise RuntimeError("Received message is not an observation!")

        return msg

    def is_ready(self):
        # Wait and consume a Ready control message
        msg = self.receive()
        if (msg.message_type == MessageType.Control.value
                and msg.step_type == StepType.Ready.value):
            return True
        else:
            return False

    def close(self):
        try:
            if self.client_socket is not None:
                try:
                    self.client_socket.shutdown(socket.SHUT_RDWR)
                except Exception:
                    pass
                self.client_socket.close()
        finally:
            if self.server_socket is not None:
                self.server_socket.close()

    # def __del__(self):
    #     # print("Destroying Messenger object.")
    #     self.close()